bbq/compiler/desugar.go

```diff
@@ -0,0 +1,353 @@
+	return ast.AcceptDeclaration[ast.Declaration](declaration, d)
+}
+
+func (d *Desugar) VisitVariableDeclaration(declaration *ast.VariableDeclaration) ast.Declaration {
+	return declaration
+}
+
+func (d *Desugar) VisitFunctionDeclaration(declaration *ast.FunctionDeclaration) ast.Declaration {
+	return declaration
+}
+
+func (d *Desugar) VisitSpecialFunctionDeclaration(declaration *ast.SpecialFunctionDeclaration) ast.Declaration {
+	return declaration
+}
+
+func (d *Desugar) VisitAttachmentDeclaration(declaration *ast.AttachmentDeclaration) ast.Declaration {
+	return declaration
+}
+
+func (d *Desugar) VisitCompositeDeclaration(declaration *ast.CompositeDeclaration) ast.Declaration {
+	existingMembers := declaration.Members.Declarations()
+	existingMemberCount := len(existingMembers)
+
+	// Recursively de-sugar nested declarations (functions, types, etc.)
+
+	desugaredMembers := make([]ast.Declaration, 0, existingMemberCount)
+	membersDesugared := false
+
+	for _, member := range existingMembers {
+		desugaredMember := d.desugarDeclaration(member)
+		membersDesugared = membersDesugared || (desugaredMember != member)
+		desugaredMembers = append(desugaredMembers, desugaredMember)
+	}
+
+	// Copy over inherited default functions.
+
+	compositeType := d.elaboration.CompositeDeclarationType(declaration)
+	inheritedDefaultFuncs := d.copyInheritedFunctions(compositeType)
+
+	// Optimization: If none of the existing members got updated or,
+	// if there are no inherited members, then return the same declaration as-is.
+	if !membersDesugared && len(inheritedDefaultFuncs) == 0 {
+		return declaration
+	}
+
+	updatedMembers := make([]ast.Declaration, existingMemberCount)
+	copy(updatedMembers, desugaredMembers)
+
+	updatedMembers = append(updatedMembers, inheritedDefaultFuncs...)
+
+	updatedDecl := ast.NewCompositeDeclaration(
+		d.memoryGauge,
+		declaration.Access,
+		declaration.CompositeKind,
+		declaration.Identifier,
+		declaration.Conformances,
+		ast.NewMembers(d.memoryGauge, updatedMembers),
+		declaration.DocString,
+		declaration.Range,
+	)
+
+	// Update elaboration. Type info is needed for later steps.
+	d.elaboration.SetCompositeDeclarationType(updatedDecl, compositeType)
+
+	return updatedDecl
+}
+
+func (d *Desugar) copyInheritedFunctions(compositeType *sema.CompositeType) []ast.Declaration {
+	directMembers := compositeType.Members
+	allMembers := compositeType.GetMembers()
+
+	inheritedMembers := make([]ast.Declaration, 0)
+
+	for memberName, resolver := range allMembers {
+		if directMembers.Contains(memberName) {
+			continue
+		}
+
+		member := resolver.Resolve(
+			nil,
+			memberName,
+			ast.EmptyRange,
+			func(err error) {
+				if err != nil {
+					panic(err)
+				}
+			},
+		)
+
+		// Inherited functions are always from interfaces
+		interfaceType := member.ContainerType.(*sema.InterfaceType)
+
+		elaboration, err := d.config.ElaborationResolver(interfaceType.Location)
```

Comment:
There is a caveat here: In order to make this work for all/complex cases, this "imported"/"brought-over" elaboration also needs to be stored (or may be merged with the current program's elaborations). Because the compiler need type-info to compile certain things, and for these copied-over functions, type information is only available in their original elaboration (not in current program's elaboration).

This needs to be handled properly (this PR does not handles this).